### PR TITLE
Partial fix for #6755. Batch three of type hints from type comments.

### LINKED ISSUE #6755: Create more Python type hints from type comments
I am always frustrated when I see Python function comments that contain type information that cannot be used by [`mypy`](https://mypy.readthedocs.io/en/stable).

### Describe the problem that you'd like solved

Use the information in the type comments to create [Python type hints](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html) that `mypy` can use to find errors in our code.
https://github.com/internetarchive/openlibrary/blob/39962157c12465d050e43458444aba9a59661c8d/openlibrary/catalog/merge/merge_marc.py#L17-L24
could be converted to
```python
def build_titles(title: str) -> dict:
    """
    Uses a full title to create normalized and short title versions.
    :param str title: Full title of an edition
    :return: An expanded dict of title variations
    """
```
Instances of these `PyCharm-style` comments can be found in 56 files in our codebase by running: `git grep -E ':param|:rtype'`  Does `PyCharm` have any [builtin capabilities](https://www.jetbrains.com/help/pycharm/type-hinting-in-product.html) to do these conversions?
*https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html
### Please do not modify all 56 files in a single pull request.
Pull requests that modify 5 to 10 files are much easier for project maintainers to review.  Lessons learned in early pull requests can improve the conversion process and accelerate subsequent conversions and reviews.

We have already used @ilevkivskyi's wonderful [`com2ann`](https://pypi.org/project/com2ann) tool to convert some type comments into Python type hints but that tool does not currently convert `:param type name:` and `:rtype:` type comments.

### Proposal & Constraints

Either manually or with some code, extract the type information from these function comments to create Python type hints.  While doing so and in code review, read the function body to ensure that the data types match the comments.

### PR DESCRIPTION
**Pull request by scottbarnes**
Partially closes #6755 

Refactor

### Technical

This batch entered the Solar system and required some use of `is not None` and `isinstance(x, y)` such that output could be changed. I... hope it has not changed.

I can see how reasonable people might disagree about whether or not some of the `:param:` comments add anything. I was trying to imagine myself in the situation of someone unfamiliar with the code base, as I am, and whether the `:param:` comments added something useful beyond the type hints and parameter names. I have no idea how much I can generalize from my own experience, if that’s even the right approach, though.

Some additional things to note:
- in `openlibrary/tests/solr/test_update_work.py`: the use of `Any` in the various functions to make make authors, editions, etc.
- in `openlibrary/solr/update_work.py`:
    - `get_last_modified()` (line 630). I wasn’t able to satisfy mypy using the comprehension format. Maybe there is a better way than a `for` loop to deal with the possibility of `datetimestr_to_int()` having None as an argument.
    - `update_author()` (line 1246). I am not sure about the approach of raising a `TypeError`. If `a` is `None`, or at least not a `dict`, an AttributeError will be raised shortly anyway I think, but this way mypy can type check it more thoroughly. I also saw the caller will log an error if an exception is raised.
- in `openlibrary/core/vendors.py`:
    - I made `resources` `Any` on line 117. After briefly consulting https://webservices.amazon.com/paapi5/documentation/get-items.html#resources-parameter it wasn’t clear to me what `resources` would be.
    - Similarly, with `serialize()` on line 155, I made `products` `Any` as I couldn’t figure out what sort of object was being returned from the API, and it looked as if I had have a reviewed and approved Amazon Associate account I could register to use the API.

### Testing

Ideally no behavior should change and mypy should pass.

### Stakeholders

@cclauss

### CODE REVIEW

**Comment by cclauss** [comment]
Please move the changes to the two solr files to a separate PR.  The file scripts/solr_builder/solr_builder/fn_to_cli.py has not been updated to use the new type annotation syntax so let’s be cautious about typing changes in solr.

**Comment by scottbarnes** [comment]
I seem to have messed this commit up by wrongly using `git --amend` and `reset --soft HEAD^`. I'll see if I can figure out how to fix that, and if not, I will delete this PR and create another.

File: openlibrary/core/vendors.py [hunk 1]
```diff
@@ -272,58 +273,60 @@ def serialize(product) -> dict:
 
 
 @public
-def get_amazon_metadata(id_: str, id_type: str = 'isbn', resources=None):
+def get_amazon_metadata(
+    id_: str, id_type: str = 'isbn', resources: Any = None
```
Comment (cclauss) [inline openlibrary/core/vendors.py:277]:
```suggestion
    id_: str, id_type: Literal['asin', 'isbn'] = 'isbn', resources: Any = None
```

File: openlibrary/core/vendors.py [hunk 2]
```diff
@@ -454,30 +450,31 @@ def cached_get_amazon_metadata(*args, **kwargs):
...
     """
-    :param str isbn: Unnormalisied ISBN10 or ISBN13
+    :param str input_isbn: Unnormalisied ISBN10 or ISBN13
     :return: Metadata for a single BWB book, currently listed on their catalog, or
              an error dict.
-    :rtype: dict or None
     """
 
-    isbn = normalize_isbn(isbn)
+    isbn = normalize_isbn(input_isbn)
```
Comment (cclauss) [inline openlibrary/core/vendors.py:460]: Why two different variables?  I don't see where this extra code buys us anything.

File: openlibrary/utils/isbn.py [hunk 3]
```diff
@@ -60,13 +61,11 @@ def isbn_10_to_isbn_13(isbn_10):
...
 
-def to_isbn_13(isbn):
+def to_isbn_13(input_isbn: str) -> str | None:
     """
     Tries to make an isbn into an isbn13; regardless of input isbn type
-    :param str isbn:
-    :rtype: str|None
     """
-    isbn = normalize_isbn(isbn)
+    isbn = normalize_isbn(input_isbn)
```
Comment (cclauss) [inline openlibrary/utils/isbn.py:68]: Same.

File: openlibrary/core/vendors.py (hunk 2 above)
Comment (scottbarnes) [reply openlibrary/core/vendors.py:460]: I realize I am probably not handling this properly, but that was my attempt to address the `isbn` parameter to `get_betterworldbooks_metadata(isbn: str)` being a string, and the output from `normalize_isbn()` being `str | None`. When the ISBN is bad on its face, even prior to the checksum, `normalize_isbn()` will return `None`.

Without this change I was getting:
```
openlibrary/core/vendors.py: note: In function "get_betterworldbooks_metadata":
openlibrary/core/vendors.py:460: error: Incompatible types in assignment (expression has type "Optional[str]", variable has type "str")  [assignment]
        isbn = normalize_isbn(isbn)
               ^
```

My other attempt was as follows:
```
    possible_isbn = normalize_isbn(isbn)
    if possible_isbn is None:
        return None
    isbn = possible_isbn
```


For context, here's `normalize_isbn()`.
```
def normalize_isbn(isbn: str) -> str | None:
    """
    Takes an isbn-like string, keeps only numbers and X/x, and returns an ISBN-like
    string or None.
    Does NOT validate length or checkdigits.
    """
    return isbn and canonical(isbn) or None
```

Comment (scottbarnes) [reply openlibrary/core/vendors.py:460]: Ah, what a good idea. Perfect. Thank you for suggesting that. I will make these changes. :)

**Comment by scottbarnes** [comment]
My apologies for the repeated pushes. I realized I had (1) created a comment that was a lie, and (2) missed some places `Literal['asin', 'isbn'] = 'isbn'` would make sense.

**Review by cclauss — APPROVED** [review]

**Comment by cclauss** [comment]
Thanks for doing this!

**Pull request merged**